engine/apps/schedules/quality.py

```diff
@@ -0,0 +1,322 @@
+            day.start,
+            working_hours_start,
+            tzinfo=pytz.UTC,
+        )
+
+        work_end = datetime.datetime.combine(
+            day.end,
+            working_hours_end,
+            tzinfo=pytz.UTC,
+        )
+
+        if (day.start < work_start and day.end < work_start) or (day.start > work_end and day.end > work_end):
+            continue
+        elif work_start <= day.start <= work_end and work_start <= day.end <= work_end:
+            result.append((day.start, day.end))
+        elif day.start < work_start and day.end > work_end:
+            result.append((work_start, work_end))
+        elif day.start >= work_start:
+            result.append((day.start, work_end))
+        else:
+            result.append((work_start, day.end))
+
+    return result
+
+
+def get_inside_working_hours_score(events: list[dict], users: dict[str, User]):
+    inside_working_hours_duration = datetime.timedelta(seconds=0)
+
+    for event in events:
+        user_pks = [user["pk"] for user in event["users"]]
+
+        for user_pk in user_pks:
+            start, end = event["start"], event["end"]
+            if event["all_day"]:
+                start = get_day_start(start)
+                end = get_day_end(end)
+
+            user = users[user_pk]
+            spans = spans_inside_working_hours(start, end, user)
+            inside_working_hours_duration += timedelta_sum(span[1] - span[0] for span in spans)
+
+    events_duration = timedelta_sum(event_duration(event) for event in events)
+
+    if events_duration:
+        score = inside_working_hours_duration / events_duration
+    else:
+        score = 1
+
+    return score
+
+
+def get_balance_score(events: list[dict]) -> float:
+    users_to_events_map = {}
+    for event in events:
+        user_pks = [user["pk"] for user in event["users"]]
+
+        for user_pk in user_pks:
+            if user_pk in users_to_events_map:
+                users_to_events_map[user_pk].append(event)
+            else:
+                users_to_events_map[user_pk] = [event]
+
+    duration_map = {}
+    for user_pk, events in users_to_events_map.items():
+        duration_map[user_pk] = timedelta_sum(event_duration(event) for event in events)
+
+    score = 0
+    number_of_pairs = 0
+    for user_1 in duration_map:
+        for user_2 in duration_map:
+            if user_1 == user_2:
+                continue
+            score += min(duration_map[user_1], duration_map[user_2]) / max(duration_map[user_1], duration_map[user_2])
+            number_of_pairs += 1
+
+    if number_of_pairs == 0:
+        balance_score = 1
+    else:
+        balance_score = score / number_of_pairs
+
+    return balance_score
+
+
+def get_balance_outside_working_hours(events: list[dict], users: dict[str, User]) -> float:
+    users_to_events_map = {}
+    for event in events:
+        user_pks = [user["pk"] for user in event["users"]]
+
+        for user_pk in user_pks:
```

Comment:
If this will be calculated for several functions, it could make sense to have a helper function at the module level?

Reply:
👍 , removed this completely